widget/infprogressbar.go

```diff
@@ -0,0 +1,170 @@
+package widget
+
+import (
+	"image/color"
+	"time"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+const infiniteRefreshRate = 50 * time.Millisecond
+
+type infProgressRenderer struct {
+	objects  []fyne.CanvasObject
+	bar      *canvas.Rectangle
+	progress *InfProgressBar
+}
+
+// MinSize calculates the minimum size of a progress bar.
+func (p *infProgressRenderer) MinSize() fyne.Size {
+	// this is to create the same size infinite progress bar as regular progress bar
+	text := textMinSize("100%", theme.TextSize(), fyne.TextStyle{})
+
+	return fyne.NewSize(text.Width+theme.Padding()*4, text.Height+theme.Padding()*2)
+}
+
+func (p *infProgressRenderer) updateBar() {
+	progressSize := p.progress.Size()
+	barWidth := p.bar.Size().Width
+	barPos := p.bar.Position()
+
+	maxBarWidth := progressSize.Width / 5  // 1/5 of bar
+	minBarWidth := progressSize.Width / 20 // 1/20 of bar
+	stepSize := (int)(progressSize.Width / 50)
+
+	// check to make sure inner bar is sized correctly
+	// if bar is on the first half of the progress bar, grow it up to (width / 5)
+	// if on the second half of the progress bar width, shrink it down until it gets to (width / 20)
+	if barWidth < maxBarWidth && barPos.X+barWidth < progressSize.Width/2 {
+		// slightly increase width
+		newBoxSize := fyne.Size{Width: barWidth + stepSize, Height: progressSize.Height}
+		p.bar.Resize(newBoxSize)
+	} else {
+		// move bar to the right by stepSize
+		barPos.X += stepSize
+
+		if barWidth <= minBarWidth {
+			// loop around to start when bar goes to end
+			barPos.X = 0
+			stepSize = 0
+			// set box size to 0
+			newBoxSize := fyne.Size{Width: minBarWidth, Height: progressSize.Height}
+			p.bar.Resize(newBoxSize)
+		} else if barPos.X+barWidth > progressSize.Width {
+			// crop to the end of the bar
+			barWidth = progressSize.Width - barPos.X
+			newBoxSize := fyne.Size{Width: barWidth, Height: progressSize.Height}
+			p.bar.Resize(newBoxSize)
+		}
+	}
+
+	p.bar.Move(barPos)
+}
+
+// Layout the components of the progress bar widget
+func (p *infProgressRenderer) Layout(size fyne.Size) {
+	// set height of progress bar
+	p.updateBar()
+}
+
+// ApplyTheme is called when the progress bar may need to update it's look
+func (p *infProgressRenderer) ApplyTheme() {
+	p.bar.FillColor = theme.PrimaryColor()
+
+	p.Refresh()
+}
+
+func (p *infProgressRenderer) BackgroundColor() color.Color {
+	return theme.ButtonColor()
+}
+
+func (p *infProgressRenderer) Refresh() {
+	p.updateBar()
+	canvas.Refresh(p.progress)
+}
+
+func (p *infProgressRenderer) Objects() []fyne.CanvasObject {
+	return p.objects
+}
+
+// InfProgressBar widget creates a horizontal panel that indicates waiting indefinitely
+// An infinite progress bar loops 0% -> 100% until Stop() is called
```

Comment:
probably wants "and back" after 100% with this new animation.

Reply:
Updated!